### PR TITLE
Trigger Automation If Customer Joins

### PR DESCRIPTION
**Pull request by TomAbel**
Ready for review
solves #32

### CODE REVIEW

File: views/templates/admin/smaily_configure.tpl [hunk 1]
```diff
@@ -181,6 +181,38 @@
               </p>
             </div>
           </div>
+          <div class="form-group">
+            <label class="control-label col-lg-2">
+              {l s="Trigger Opt-in If Customer Joins With Newsletter Subscription" mod='smailyforprestashop'}
```
Comment (kaittodesk) [inline views/templates/admin/smaily_configure.tpl:186]: Improve phrasing. Use something like:

> Trigger opt-in on customer sign-up

And add help block under the slide button (like here - https://github.com/sendsmaily/smaily-prestashop-module/blob/master/views/templates/admin/smaily_configure.tpl#L157-L159):

> Note! Opt-in will only be triggered when customer signs-up for newsletter. Changes to newsletter subscription in admin won't trigger an opt-in.

File: views/templates/admin/smaily_configure.tpl [hunk 2]
```diff
@@ -181,6 +181,38 @@
...
+                <label for="Enable">{l s="Yes" mod='smailyforprestashop'}</label>
+                <input type="radio" name="SMAILY_ENABLE_OPTIN_IF_CUSTOMER_JOINS_WITH_SUBSCRIPTION" id="Disable" value="0" {($smaily_enable_optin_if_customer_joins_with_subscription == 0) ? 'checked' :'' }>
+                <label for="Disable">{l s="No" mod='smailyforprestashop'}</label>
+                <a class="slide-button btn"></a>
+              </span>
+            </div>
+          </div>
+          <div class="form-group">
+            <label class="control-label col-lg-2">
+              {l s="Opt-in Automation If Customer Joins With Newsletter Subscription" mod='smailyforprestashop'}
```
Comment (kaittodesk) [inline views/templates/admin/smaily_configure.tpl:200]: Improve phrasing:

> Automation to trigger on customer sign-up

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: views/templates/admin/smaily_configure.tpl [hunk 3]
```diff
@@ -181,6 +181,38 @@
...
+            </label>
+            <div class="col-lg-10">
+              <select name="SMAILY_CUSTOMER_JOIN_AUTORESPONDER" id="SMAILY_CUSTOMER_JOIN_AUTORESPONDER">
+                {if empty($smaily_customer_join_autoresponder)}
+                <option value="">
+                  {l s="No Autoresponder" mod='smailyforprestashop'}
+                </option>
+                {else}
+                <option value='{$smaily_customer_join_autoresponder|json_encode}'>
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl:209]: Since autoresponders are fetched only one time, when credentials are validated, how should we display the autoresponder title?

File: smailyforprestashop.php [hunk 4]
```diff
@@ -78,12 +78,16 @@ public function install()
...
+            !Configuration::updateValue('SMAILY_CUSTOMER_JOIN_AUTORESPONDER', '') ||
             // Add tab to sidebar
             !$this->installTab('AdminAdmin', 'AdminSmailyforprestashopAjax', 'Smaily for PrestaShop') ||
             // Add Newsletter subscription form.
             !$this->registerHook('footerBefore') ||
             !$this->registerHook('leftColumn') ||
-            !$this->registerHook('rightColumn')
+            !$this->registerHook('rightColumn') ||
+            // User has option to trigger opt-in when customer joins store & newsletter through sign-up.
+            !$this->registerHook('actionCustomerAccountAdd')
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply smailyforprestashop.php:90]: Fixed this with upgrade script.

File: smailyforprestashop.php [hunk 5]
```diff
@@ -78,12 +78,16 @@ public function install()
             !Configuration::updateValue('SMAILY_RSS_LIMIT', '50') ||
             !Configuration::updateValue('SMAILY_RSS_SORT_BY', 'date_upd') ||
             !Configuration::updateValue('SMAILY_RSS_SORT_ORDER', 'desc') ||
+            !Configuration::updateValue('SMAILY_ENABLE_OPTIN_IF_CUSTOMER_JOINS_WITH_SUBSCRIPTION', 0) ||
+            !Configuration::updateValue('SMAILY_CUSTOMER_JOIN_AUTORESPONDER', '') ||
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply smailyforprestashop.php]: Fixed this with upgrade script.

File: smailyforprestashop.php [hunk 6]
```diff
@@ -194,6 +201,9 @@ public function getContent()
                     $escaped_sync_additional[] = pSQL($value);
                 }
             }
+            $enable_optin_if_customer_joins_with_subscription = pSQL(Tools::getValue('SMAILY_ENABLE_OPTIN_IF_CUSTOMER_JOINS_WITH_SUBSCRIPTION'));
+            $customer_join_autoresponder = pSQL((Tools::getValue('SMAILY_CUSTOMER_JOIN_AUTORESPONDER')));
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply smailyforprestashop.php]: Renamed.

File: views/templates/admin/smaily_configure.tpl (hunk 1 above)
Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl]: Renamed and added help block

File: views/templates/admin/smaily_configure.tpl (hunk 2 above)
Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl]: Replaced title

File: views/templates/admin/smaily_configure.tpl (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaittodesk) [reply views/templates/admin/smaily_configure.tpl]: You wouldn't need to display the autoresponder title separately. You could just add the `selected` attribute to option with value of `SMAILY_OPTIN_AUTORESPONDER`.

You could achieve this by reworking how `ajaxProcessGetAutoresponders` method works.

Instead of returning autoresponders as:
```json
{
  "success": true,
  "autoresponders": [
    {"id":14,"trigger_type":"form_submitted","title":"GOTO","is_enabled":false},
    {"id":15,"trigger_type":"form_submitted","title":"GOTO without loop","is_enabled":false}
  ]
}
```

You could do:
```json
{
  "success": true,
  "autoresponders": [
    {"id":14,"title":"GOTO","is_selected":false},
    {"id":15,"title":"GOTO without loop","is_selected":true}
  ]
}
```

And when you need to define by which configuration option a autoresponder should be selected by, you could provide a data field (or a query argument) in the query.

Though this is kind of a hackish approach.

An alternative (and a better solution) would be to use Javascript in controlling which option gets selected. Although that would mean that you would need to provide some sort of settings object to Javascript.

A simplest way to achieve that is to make a separate AJAX request to fetch some of the settings and use these values when rendering autoresponder options. An important thing to note here is that options rendering should only happen after first request has succeeded, as AJAX requests are always asynchronous you cannot guarantee that settings have been loaded before autoresponder options rendering has begun.

If you have any questions regarding the proposed solutions or there is something unclear, don't hesitate to ask.

[the remark this answers is not in the document]
Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl]: Only saving autoresponder ID to database. Also noticed user doesn't have a choice to select no autoresponder, added a default empty option to optin autoresponder list.
Used the same method for abandoned cart autoresponders aswell.

File: upgrade/Upgrade-1.5.0.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+/**
+ * Migrates existing abandoned cart fields to new structure.
+ */
+function upgrade_module_1_5_0($object)
+{
+    return (Configuration::updateValue('SMAILY_OPTIN_ENABLED', 0) &&
```
Comment (kaittodesk) [inline upgrade/Upgrade-1.5.0.php:34]: Add `SMAILY_CART_AUTORESPONDER` value conversion to upgrade. Older versions of the module store JSON object in the configuration value, and it shouldn't break existing installations.

File: views/templates/admin/smaily_configure.tpl [hunk 8]
```diff
@@ -181,6 +181,35 @@
...
+            </div>
+          </div>
+          <div class="form-group">
+            <label class="control-label col-lg-2">
+              {l s="Automation to trigger on customer sign-up" mod='smailyforprestashop'}
+            </label>
+            <div class="col-lg-10">
+              <select name="SMAILY_OPTIN_AUTORESPONDER" id="SMAILY_OPTIN_AUTORESPONDER" data-selected-id={$smaily_optin_autoresponder}>
+                <option value="">
+                  {l s="No Autoresponder" mod='smailyforprestashop'}
```
Comment (kaittodesk) [inline views/templates/admin/smaily_configure.tpl:208]: Perhaps it would be better to use a more explanatory phrase here. So instead of "No Autoresponder" it should say "Trigger any opt-in automation".

**Review by kaittodesk — CHANGES_REQUESTED** [review]

File: views/templates/admin/smaily_configure.tpl (hunk 8 above)
Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl:208]: Maybe it would be even better to use the same solution like in abandoned cart autoresponder?
```
<option value="">
    {l s="Select Autoresponder" mod='smailyforprestashop'}
</option>
```

File: smailyforprestashop.php [hunk 9]
```diff
@@ -344,12 +345,45 @@ public function getContent()
...
+    /**
+     * Unserialize autoresponder to array for template.
+     *
+     * @param string $autoresponder a:2:{s:4:\"name\";s:14:\"Newsletter\";s:2:\"id\";i:28;}
+     * @return array $autoresponder {["name"]=> string(14) "Newsletter" ["id"]=> int(28)}
+     */
+    private function unSerializeAutoresponderForTemplate($autoresponder) {
+        $autoresponder = str_replace('\"', '"', $autoresponder);
+        return unserialize($autoresponder);
+    }
```
[the remark this answers is not in the document]
Comment (TomAbel) [reply smailyforprestashop.php]: Removed

File: views/templates/admin/smaily_configure.tpl (hunk 8 above)
Comment (kaittodesk) [reply views/templates/admin/smaily_configure.tpl:208]: Yes, if we would have Abandoned Cart trigger, but unfortunately we do not. Abandoned Cart shouldn't trigger opt-in automations. Maybe we should even make Abandoned Cart autoresponder selection mandatory?

Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl:208]: I mean we're using autoresponder instead of automation on every plugin, only using automation in documentation. I am suggesting we keep with the consistency.

Comment (TomAbel) [reply views/templates/admin/smaily_configure.tpl:208]: > Maybe we should even make Abandoned Cart autoresponder selection mandatory?
I think we should make it  mandatory

File: upgrade/Upgrade-1.5.0.php (hunk 7 above)
Comment (TomAbel) [reply upgrade/Upgrade-1.5.0.php]: Added conversion to upgrade script

File: views/templates/admin/smaily_configure.tpl (hunk 8 above)
Comment (kaittodesk) [reply views/templates/admin/smaily_configure.tpl:208]: Oh, OK. Yeah, we should use the same terms everywhere. Change the term to "Trigger any opt-in autoresponder" then (omitting "opt-in" might make user think that any autoresponder is triggered, no matter what the trigger is). Most likely it will change in the future to "automation", but at least it is consistent for now.

**Review by kaittodesk — APPROVED** [review]

File: translations/et.php [hunk 10]
```diff
@@ -43,6 +43,10 @@
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_bb7b3c6261b41de1bd1fefa377a75f43'] = 'Identifikaator on vajalik turvalisuse tagamiseks. Kasuta automaatselt loodud identifikaatorit või asenda enda valitud väärtusega. ';
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_5588ea18ef47155105d7d11abcb7e4e1'] = 'Croni URL';
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_9cc2022eb76bc383c713c406b6ae96f2'] = 'Et aktiveerida automaatne kontaktide sünkroniseerimine seadistage Cron oma teenusepakkuja juures.';
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_daf2a96781f203ddd0e0c8ba8fa25a2f'] = 'Käivita opt-in kasutaja registreerimisel.';
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_dc7361b8c5d6e2b9f95b8968533d7b14'] = 'Opt-in käivitatakse ainult siis, kui kasutaja loob konto & liitub uudiskirjaga. Muutused uudiskirja sätetele adminpaneelis ei käivita opt-in\'ii.';
```
Comment (kaittodesk) [inline translations/et.php:47]: You have one extra "i" in the end and I would change the ampersand to `ja`:

```suggestion
$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_dc7361b8c5d6e2b9f95b8968533d7b14'] = 'Opt-in käivitatakse ainult siis, kui kasutaja loob konto ja liitub uudiskirjaga. Muutused uudiskirja sätetele adminpaneelis ei käivita opt-in\'i.';
```

File: translations/et.php [hunk 11]
```diff
@@ -43,6 +43,10 @@
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_bb7b3c6261b41de1bd1fefa377a75f43'] = 'Identifikaator on vajalik turvalisuse tagamiseks. Kasuta automaatselt loodud identifikaatorit või asenda enda valitud väärtusega. ';
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_5588ea18ef47155105d7d11abcb7e4e1'] = 'Croni URL';
 $_MODULE['<{smailyforprestashop}prestashop>smaily_configure_9cc2022eb76bc383c713c406b6ae96f2'] = 'Et aktiveerida automaatne kontaktide sünkroniseerimine seadistage Cron oma teenusepakkuja juures.';
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_daf2a96781f203ddd0e0c8ba8fa25a2f'] = 'Käivita opt-in kasutaja registreerimisel.';
```
Comment (kaittodesk) [inline translations/et.php:46]: I would remove the dot from the end as it will be used as the label of the form field.

File: translations/et.php [hunk 12]
```diff
@@ -43,6 +43,10 @@
...
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_dc7361b8c5d6e2b9f95b8968533d7b14'] = 'Opt-in käivitatakse ainult siis, kui kasutaja loob konto & liitub uudiskirjaga. Muutused uudiskirja sätetele adminpaneelis ei käivita opt-in\'ii.';
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_f31fe73b24620e23d03fd55013e1bf87'] = 'Kasutaja registreerimisel käivitatav automaatika.';
```
Comment (kaittodesk) [inline translations/et.php:48]: Remove the dot from the end, as this will be the name of the select option.

File: translations/et.php [hunk 13]
```diff
@@ -43,6 +43,10 @@
...
+$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_faa5e8094afca015fa4b4a702ed57645'] = 'Käivita mõni automaatika.';
```
Comment (kaittodesk) [inline translations/et.php:49]: Rephrase this to:

```suggestion
$_MODULE['<{smailyforprestashop}prestashop>smaily_configure_faa5e8094afca015fa4b4a702ed57645'] = 'Käivita iga opt-in automaatika';
```

**Review by kaittodesk — CHANGES_REQUESTED** [review]

**Review by kaittodesk — APPROVED** [review]
